File: 1802-maximum-value-at-a-given-index-in-a-bounded-array/1802-maximum-value-at-a-given-index-in-a-bounded-array.py
class Solution:
    def maxValue(self, n: int, index: int, maxSum: int) -> int:
        def check(a):
            b = max(a - index, 0)
            res = (a + b) * (a - b + 1) / 2
            b = max(a - (n - 1 - index), 0)
            res += (a + b) * (a - b + 1) / 2
            return res - a
        
        maxSum -= n
        left, right = 0, maxSum
        while (left <= right):
            mid = left + (right - left) // 2
            if check(mid) <= maxSum:
                left = mid + 1
            else:
                right = mid - 1
        return left

File: 1802-maximum-value-at-a-given-index-in-a-bounded-array/1802-maximum-value-at-a-given-index-in-a-bounded-array.cpp
class Solution {
public:
    int maxValue(int n, int index, int maxSum) {
        maxSum -= n;
        int left = 0, right = maxSum;
        while (left <= right){
            int mid = left + (right - left) / 2;
            if (check(n, index, mid) <= maxSum) {
                left = mid + 1;
            }else {
                right = mid - 1;
            }
        }
        return left;
    };
private:
    long check(int n, int index, int a){
        int b = max(a - index, 0);
        long res = long(a + b) * (a - b + 1) / 2;
        b = max(a - (n - 1 - index), 0);
        res += long(a + b) * (a - b + 1) / 2;
        return res - a;
    }
};